src/shared/loop-util.c

```diff
@@ -702,6 +738,8 @@ LoopDevice* loop_device_unref(LoopDevice *d) {
         if (!d)
                 return NULL;
 
+        d->lock_fd = safe_close(d->lock_fd);
```

Comment:
I think this should be moved to the end of this function, at least, after removing the partitions.

Reply:
Hm, how about to close `d->fd` first if `d->lock_fd >= 0`?? Then, there should be only fd we have.
IIUC `block_device_remove_all_partitions()` and `LOOP_CLR_FD` also works fine for `lock_fd`, right?
```c
        if (d->fd >= 0) {
                /* Implicitly sync the device, since otherwise in-flight blocks might not get written */
                if (fsync(d->fd) < 0)
                        log_debug_errno(errno, "Failed to sync loop block device, ignoring: %m");
                if (d->nr >= 0 && !d->relinquished) {
                        /* We are supposed to clear the loopback device. Let's do this synchronously: lock
                         * the device, manually remove all partitions and then clear it. This should ensure
                         * udev doesn't concurrently access the devices, and we can be reasonably sure that
                         * once we are done here the device is cleared and all its partition children
                         * removed. */

                        if (d->lock_fd >= 0) {
                                safe_close(d->fd);
                                d->fd = TAKE_FD(d->lock_fd);
                        } else if (flock(d->fd, LOCK_EX) < 0)
                                log_debug_errno(errno, "Failed to lock loop block device, ignoring: %m");

                        r = block_device_remove_all_partitions(d->fd);
                        if (r < 0)
                                log_debug_errno(r, "Failed to remove partitions of loopback block device, ignoring: %m");
                        if (ioctl(d->fd, LOOP_CLR_FD) < 0)
                                log_debug_errno(errno, "Failed to clear loop device, ignoring: %m");
                }
                safe_close(d->fd);
        }
```

Reply:
We can't because of the mentioned ABBA issue:

1. We take a lock on /dev/loop-control
2. We take a lock on the actual loopback block device
3. We then remove all partitions + issue LOOP_CLR_FD
4. We release both locks

If we do things that way, and allocation works like this:

1. We take a lock on /dev/loop-control
2. We ask the control device for next available loopback device, and open and lock it
3. We delete any remaining partitions on it
4. We release the /dev/loop-control device, and possibly the loopback device lock

Then we safely avoided the race that between asking which device is available next, and actually allocating it and setting it up some other participant also takes it and then maybe releases it, or is still in progress of releasing it after having used it.

But this requires us to follow a strict locking protocol: to avoid ABBA deadlocks, we always must acquire /dev/loop-control first, and only the block device. Now, when loop_device_unref() is entered there might be a lock in effect on the loopback device, but there is not on /dev/loop-control. But we want one on that, and given the lock order constraints we must thus must give up the block device lock first, then acquire the control lock and only then acquire the block device lock again.

if we don't do it that way (and instead do as you suggest), we might deadlock: because a current owner which locked the block device might ask for the /dev/loop-control lock, while another process might just try to allocate it, thus already acquire /dev/loop-control and then goes on tries to lock the block device. Now each of the two processes has a lock on one device and tries to acquire one on the other, but can't because the other one has it. Deadlock...

So we fix this by giving up the lock temporarily here, and re-acquire it in the order our lock protocol dictates: get the /dev/loop-control lock first, then acquire the loopback block device. And never the other way round.

Reply:
Hm, interesting. Thank you for the explanation.